### ISSUE TITLE
cargo test --doc does not respect CARGO_BUILD_JOBS env variable

### ISSUE THREAD

**Issue by tmpolaczyk** [rust-lang/cargo#10702]
### Problem

I use a global variable CARGO_BUILD_JOBS=4 because the default configuration results in my machine running out of memory when compiling some rust projects. However recently I noticed that this fails when there are many doc tests in the project, because it looks like all the doc tests are compiled in parallel.

### Steps

1. Find some project with many doc tests
2. export CARGO_BUILD_JOBS=1
3. cargo test --doc
4. In a separate terminal, run `pidof rustc` and notice how there are as multiple processes, instead of the expected 1.
5. This results in the tests failing, in my case because the linker is killed because the system has run out of memory.

### Possible Solution(s)

In my opinion the correct behavior is to respect CARGO_BUILD_JOBS when compiling the tests.

As a workaround, you can use

    cargo test -- --test-threads 1

to limit the number of build jobs to 1, however this also affects normal tests, so running tests will be much slower.

### Version

```text
cargo 1.61.0 (a028ae4 2022-04-29)
release: 1.61.0
commit-hash: a028ae42fc1376571de836be702e840ca8e060c2
commit-date: 2022-04-29
host: x86_64-unknown-linux-gnu
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.80.0-DEV (sys:0.4.51+curl-7.80.0 vendored ssl:OpenSSL/1.1.1m)
os: Ubuntu 22.04 (jammy) [64-bit]
```

**Comment by weihanglo** [comment]
To my understanding, `rustdoc` collects each doctest and compiles them by invoking `rustc` on-the-fly[^1], and that is a part of **test fn**[^2], which will be executed by libtest. 

That tells us why `build.jobs` (controls cargo-rustc jobserver) doesn't affect parallelism of doctest, but `--test-threads` does (controls parallelism of libtest).

[^1]: https://github.com/rust-lang/rust/blob/9e26dc71fdf370ced4d7b9971c5828bef9162290/src/librustdoc/doctest.rs#L323-L327
[^2]: https://github.com/rust-lang/rust/blob/9e26dc71fdf370ced4d7b9971c5828bef9162290/src/librustdoc/doctest.rs#L1033-L1037

**Comment by weihanglo** [comment]
I would recommend opening an issue against rust-lang/rust, though I don't have any viable fix for `rustdoc` in mind at this moment 😞. In the meanwhile, I suggest you to stick with `cargo test --doc -- --test-threads 1` in order not to affect unit/integration tests.

Closing as the behaviour is controlled by `rustdoc`. Cargo gets no knowledge about how a test fn being executed. If you think this issue shouldn't be closed, feel free to reopen and discuss further. Thanks :)

**Comment by Urgau** [comment]
@weihanglo I disagree with your assessment, `rustdoc` has no knowledge of `cargo` and it's environment variables and shouldn't have any.

If cargo want to control the number of test parallelism with it's env `CARGO_BUILD_JOBS` it should pass the right command line argument to `rustdoc`.

I would also argue that `CARGO_BUILD_JOBS` shouldn't control this at all. Yes, technically `rustdoc` build something but it's irrelevant, it's testing: testing isn't the same as building.

**Comment by weihanglo** [comment]
Hmm… Sorry about the misleading. I was not suggesting that rustc should recognize `CARGO_BUILD_JOBS` and I totally agreed `CARGO_BUILD_JOBS` shouldn't touch how tests are executed. What I wanted to say is that if they want to let Cargo control those rustc invocations from rustdoc tests, they should ask rustdoc for providing such an interface (though that's hard and might be a breaking change).

For now, I feel like it's even not a problem since there is already a alternative solution (`cargo test --doc -- --test-threads 1`). I shouldn't head them towards opening another issue in rust-lang/rust. @Urgau, sorry for the mistake and not saying it clearer. 😞

**Comment by tmpolaczyk** [comment]
Hi @weihanglo, unfortunately that workaround doesn't work, because the tests are compiled again when running the second command. I also tried variations like using `--all-features` in both commands, but that still causes a crash in the second command.

    cargo test --doc --all-features -- --test-threads 1
    cargo test --all-features

In my case, the root cause of this issue is that the default configuration results in an out of memory crash. I can fix that by setting the CARGO_BUILD_JOBS variable globally to a lower number of threads. And that fixes the problem everywhere except when running doc tests. So while I understand that because of some implementation details my desired behavior of respecting the CARGO_BUILD_JOBS variable cannot be implemented, I strongly believe that it would be the correct behavior. Because my machine simply cannot handle the full parallelism of rustc, so I would expect a single environment variable to fix that issue globally. Because even if there is a workaround like running two separate cargo test commands, every other person that has a similar problem will also need to run that sequence of commands.

**Comment by weihanglo** [comment]
Ok, I see. Whether that there is a change in rustdoc, cargo still needs to integrate with that or implement something on its own. I'll leave this open for tracking any future possibilities.

**Comment by tmpolaczyk** [comment]
Hi, I don't know if this has been fixed already, but I saw that the cargo parallel frontend solved this exact same problem using the "jobserver protocol":

> You might wonder how interprocess parallelism and intraprocess parallelism interact. If we have 20 parallel rustc invocations and each one can have up to 16 threads running, could we end up with hundreds of threads on a machine with only tens of cores, resulting in inefficient execution as the OS tries its best to schedule them?
>
> Fortunately no. The compiler uses the [jobserver protocol](https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html) to limit the number of threads it creates. If a lot of interprocess parallelism is occuring, intraprocess parallelism will be limited appropriately, and the number of threads will not exceed the number of cores.

Source: <https://blog.rust-lang.org/2023/11/09/parallel-rustc.html#putting-it-all-together>
